### PR TITLE
Initial revision of SHA-3

### PR DESCRIPTION
**Pull request by SparkiDev**

### CODE REVIEW

File: wolfssl/wolfcrypt/sha3.h [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+
+/* code submitted by raphael.huck@efixo.com */
```
Comment (toddouska) [inline wolfssl/wolfcrypt/sha3.h]: This doesn't seem correct.

File: wolfcrypt/src/sha3.c [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+{
+    return wc_Sha3Final(sha3, hash, SHA3_256_COUNT, SHA3_256_DIGEST_SIZE);
+}
+
+WOLFSSL_API void wc_Sha3_256_Free(Sha3* sha3)
+{
+    wc_Sha3Free(sha3);
+}
+
+WOLFSSL_API int wc_Sha3_256_GetHash(Sha3* sha3, byte* hash)
```
Comment (toddouska) [inline wolfcrypt/src/sha3.c]: We should document the difference between GetHash and Final somewhere.

File: wolfcrypt/src/sha3.c [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+{
+    return wc_Sha3GetHash(sha3, hash, SHA3_256_COUNT, SHA3_256_DIGEST_SIZE);
+}
+
+WOLFSSL_API int wc_Sha3_256_Copy(Sha3* src, Sha3* dst)
```
Comment (toddouska) [inline wolfcrypt/src/sha3.c]: Copy should be documented as well.  Does Init for example need to be called on destination first or is that allowed?

File: wolfcrypt/src/sha3.c [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+{
+    return wc_Sha3Update(sha3, data, len, SHA3_256_COUNT);
+}
+
+WOLFSSL_API int wc_Sha3_256_Final(Sha3* sha3, byte* hash)
+{
+    return wc_Sha3Final(sha3, hash, SHA3_256_COUNT, SHA3_256_DIGEST_SIZE);
+}
+
+WOLFSSL_API void wc_Sha3_256_Free(Sha3* sha3)
```
Comment (toddouska) [inline wolfcrypt/src/sha3.c]: Is Free always required?

Comment (SparkiDev) [inline wolfcrypt/src/sha3.c]: No. The async work added a free to all digest algorithms and I'm following the pattern.

File: wolfssl/wolfcrypt/sha3.h (hunk 1 above)
Comment (SparkiDev) [inline wolfssl/wolfcrypt/sha3.h]: Removed!

File: wolfcrypt/src/sha3.c (hunk 2 above)
Comment (SparkiDev) [inline wolfcrypt/src/sha3.c]: Added block comment to each function.

**Comment by SparkiDev** [comment]
Rebased

**Pull request merged**